molgenis-security/src/test/java/org/molgenis/security/twofactor/service/RecoveryServiceImplTest.java

```diff
@@ -0,0 +1,141 @@
+package org.molgenis.security.twofactor.service;
+
+import org.molgenis.data.DataService;
+import org.molgenis.data.populate.IdGenerator;
+import org.molgenis.data.populate.IdGeneratorImpl;
+import org.molgenis.data.support.DataServiceImpl;
+import org.molgenis.data.support.QueryImpl;
+import org.molgenis.security.twofactor.meta.*;
+import org.molgenis.security.user.UserService;
+import org.molgenis.security.user.UserServiceImpl;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.security.test.context.support.WithSecurityContextTestExecutionListener;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.TestExecutionListeners;
+import org.springframework.test.context.testng.AbstractTestNGSpringContextTests;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static org.mockito.Mockito.*;
+import static org.molgenis.security.core.runas.RunAsSystemAspect.runAsSystem;
+import static org.molgenis.security.twofactor.meta.RecoveryCodeMetadata.*;
+import static org.testng.Assert.assertEquals;
+
+/**
+ * <p>Used for generating and redeeming recovery codes</p>
+ */
+@ContextConfiguration(classes = { RecoveryServiceImplTest.Config.class })
+@TestExecutionListeners(listeners = WithSecurityContextTestExecutionListener.class)
```

Comment:
Can't use the annotation because of the overridden ExecutionListeners in the AbstractNGSpringContext.class. Left it the way it was. 

You can use "return mock(DataServiceImpl.class, RETURNS_DEEP_STUBS);" to mock chained methods.